compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmBuiltinOptimizationLowering.kt

```diff
@@ -102,6 +101,22 @@ class JvmBuiltinOptimizationLowering(val context: JvmBackendContext) : FileLower
                 expression.branches.removeIf() {
                     it.condition.isFalseConst() && isCompilerGenerated
                 }
+                if (expression.origin == IrStatementOrigin.ANDAND) {
+                    assert(expression.type.isBoolean()
```

Comment:
Please always add messages with additional information to asserts, including a description of what's wrong and any relevant variables (e.g. `expression.dump()` here)